gensim/models/term_similarity.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2018 Vit Novotny <witiko@mail.muni.cz>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+"""
+This module provides classes that deal with term similarities.
+"""
+
+from itertools import chain
+import logging
+from math import sqrt
+
+import numpy as np
+from scipy import sparse
+
+from gensim.matutils import corpus2csc
+from gensim.utils import SaveLoad, is_corpus
+
+logger = logging.getLogger(__name__)
+
+
+class TermSimilarityIndex(SaveLoad):
+    """
+    Retrieves most similar terms for a given term.
+
+    See Also
+    --------
+    :class:`~gensim.models.term_similarity.SparseTermSimilarityMatrix`
+        Build a term similarity matrix and compute the Soft Cosine Measure.
+
+    """
+    def most_similar(self, term, topn=10):
+        """Get most similar terms for a given term.
+
+        Return most similar terms for a given term along with the similarities.
+
+        Parameters
+        ----------
+        term : str
+            Tne term for which we are retrieving `topn` most similar terms.
+        topn : int, optional
+            The maximum number of most similar terms to `term` that will be retrieved.
+
+        Returns
+        -------
+        iterable of (str, float)
+            Most similar terms along with their similarities to `term`. Only terms distinct `term`
+            must be returned.
+
+        """
+        raise NotImplementedError
+
+
+class UniformTermSimilarityIndex(TermSimilarityIndex):
+    """
+    Retrieves most similar terms for a given term under the hypothesis that the similarities between
+    distinct terms are uniform.
+
+    Parameters
+    ----------
+    dictionary : :class:`~gensim.corpora.dictionary.Dictionary`
+        A dictionary that specifies the considered terms.
+    term_similarity : float, optional
+        The uniform similarity between distinct terms.
+
+    See Also
+    --------
+    :class:`~gensim.models.term_similarity.SparseTermSimilarityMatrix`
+        Build a term similarity matrix and compute the Soft Cosine Measure.
+
+    Notes
+    -----
+    This class is mainly intended for testing SparseTermSimilarityMatrix and other classes that
+    depend on the TermSimilarityIndex.
+
+    """
+    def __init__(self, dictionary, term_similarity=0.5):
+        self.dictionary = dictionary
+        self.term_similarity = term_similarity
+
+    def most_similar(self, t1, topn=10):
```

Comment:
Looks like this isn't ready, am I right?

Reply:
This is quite ready, just not very useful. `UniformTermSimilarityIndex` assigns a constant similarity to any pair of distinct words; the main use is for testing `SparseTermSimilarityMatrix`.

Reply:
It is also quite useful for benchmarking the maximum throughput of the `SparseTermSimilarityMatrix` constructor.

Reply:
Another use would be the construction of an identity term similarity matrix by setting `term_similarity` to zero.